spdmlib/src/protocol/algo.rs

```diff
@@ -1185,70 +1116,240 @@ impl Default for SpdmPskHintStruct {
         }
     }
 }
-
 impl AsRef<[u8]> for SpdmPskHintStruct {
     fn as_ref(&self) -> &[u8] {
         &self.data[0..(self.data_size as usize)]
     }
 }
 
-#[derive(Debug, Clone, Zeroize, ZeroizeOnDrop)]
-pub struct SpdmAeadKeyStruct {
-    pub data_size: u16,
-    pub data: Box<[u8; SPDM_MAX_AEAD_KEY_SIZE]>,
+macro_rules! create_sensitive_datatype {
+    (Name: $name:ident, Size: $size:expr) => {
+        #[derive(Debug, Clone, Zeroize, ZeroizeOnDrop)]
+        pub struct $name {
+            pub data_size: u16,
+            pub data: Box<[u8; $size]>,
+        }
+
+        impl Default for $name {
+            fn default() -> $name {
+                $name {
+                    data_size: 0,
+                    data: Box::new([0u8; $size]),
+                }
+            }
+        }
+
+        impl AsRef<[u8]> for $name {
+            fn as_ref(&self) -> &[u8] {
+                &self.data[0..(self.data_size as usize)]
+            }
+        }
+
+        impl From<BytesMutStrubbed> for $name {
+            fn from(value: BytesMutStrubbed) -> Self {
+                assert!(value.as_ref().len() <= $size);
+                let data_size = value.as_ref().len() as u16;
+                let mut data = Box::new([0u8; $size]);
+                data[0..value.as_ref().len()].copy_from_slice(value.as_ref());
+                Self { data_size, data }
+            }
+        }
+
+        impl From<&[u8]> for $name {
+            fn from(value: &[u8]) -> Self {
+                assert!(value.len() <= $size);
+                let data_size = value.len() as u16;
+                let mut data = Box::new([0u8; $size]);
+                data[0..value.len()].copy_from_slice(value.as_ref());
+                Self { data_size, data }
+            }
+        }
+    };
+}
+
+create_sensitive_datatype!(Name: SpdmDigestStruct, Size: SPDM_MAX_HASH_SIZE);
+create_sensitive_datatype!(Name: SpdmDheFinalKeyStruct, Size: SPDM_MAX_DHE_KEY_SIZE);
+create_sensitive_datatype!(Name: SpdmHandshakeSecretStruct, Size: SPDM_MAX_HASH_SIZE);
+create_sensitive_datatype!(
+    Name: SpdmDirectionHandshakeSecretStruct,
+    Size: SPDM_MAX_HASH_SIZE
+);
+create_sensitive_datatype!(Name: SpdmFinishedKeyStruct, Size: SPDM_MAX_HASH_SIZE);
+create_sensitive_datatype!(Name: SpdmMasterSecretStruct, Size: SPDM_MAX_HASH_SIZE);
+create_sensitive_datatype!(
+    Name: SpdmDirectionDataSecretStruct,
+    Size: SPDM_MAX_HASH_SIZE
+);
+create_sensitive_datatype!(Name: SpdmAeadKeyStruct, Size: SPDM_MAX_AEAD_KEY_SIZE);
+create_sensitive_datatype!(Name: SpdmAeadIvStruct, Size: SPDM_MAX_AEAD_IV_SIZE);
+create_sensitive_datatype!(Name: SpdmExportMasterSecretStruct, Size: SPDM_MAX_HASH_SIZE);
+create_sensitive_datatype!(Name: SpdmGeneralPaddingStruct, Size: SPDM_MAX_HASH_SIZE);
```

Comment:
The SpdmGeneralPaddingStruct is not so Generic, because it still uses SPDM_MAX_HASH_SIZE. Why it is called Generic?

Reply:
> The SpdmGeneralPaddingStruct is not so Generic, because it still uses SPDM_MAX_HASH_SIZE. Why it is called Generic?

the initial idea is to represent the zere_filled slice. And I changed the type SpdmGeneralPaddingStruct to SpdmZeroFilledStruct in latest commit.

Reply:
I am not sure if that is right way.
SpdmZeroFilledStruct uses SPDM_MAX_HASH_SIZE, but the IMK may be SPDM_MAX_DHE_KEY_SIZE.

I dont understand why we have to include SpdmZeroFilledStruct. Can we just remove it?

Reply:
> I am not sure if that is right way. SpdmZeroFilledStruct uses SPDM_MAX_HASH_SIZE, but the IMK may be SPDM_MAX_DHE_KEY_SIZE.
> 
> I dont understand why we have to include SpdmZeroFilledStruct. Can we just remove it?

SpdmZeroFilledStruct  is used to derive master secret.